plugins/raiseStat/raiseStat.pl

```diff
@@ -0,0 +1,185 @@
+
+################################################################
+#  Hooks used to activate the plugin during initialization
+#  and on config change events.
+my $base_hooks = Plugins::addHooks(
+	['start3',        \&checkConfig],
+	['postloadfiles', \&checkConfig],
+    ['configModify',  \&on_configModify]
+   );
+
+my @stats_to_add;
+my $active_hooks;
+my $adding_hook;
+my $next_stat;
+my $status;
+my $timeout = { time => 0, timeout => 1 };
+
+use constant {
+	INACTIVE => 0,
+	ACTIVE => 1,
+	ADDING => 2
+};
+
+sub on_unload {
+   Plugins::delHook($base_hooks);
+   changeStatus(INACTIVE);
+   message "raiseStat plugin unloading or reloading\n", 'success';
+}
+
+################################################################
+#  changeStatus() is the function responsible for adding
+#  and deleting hooks when changing plugin status.
+#  During status == 0 the plugin will be inactive and won't
+#  look for opportunities to raise stats.
+#  During status == 1 the plugin will be active and will
+#  have 'speculative' hooks added to try to look for
+#  opportunities to raise stats.
+#  During status == 2 the plugin will be active and will
+#  have 'AI_pre' hook active, on each AI call the plugin
+#  will try to raise stats if possible.
+sub changeStatus {
+	my $new_status = shift;
+	Plugins::delHook($active_hooks) if ($status == ACTIVE);
+	Plugins::delHook($adding_hook) if ($status == ADDING);
+	if ($new_status == INACTIVE) {
+		undef $next_stat;
+		undef @stats_to_add;
+	} elsif ($new_status == ACTIVE) {
+		$active_hooks = Plugins::addHooks(
+			['map_loaded',           \&on_possible_raise_chance],
+			['packet/sendMapLoaded', \&on_possible_raise_chance],
+			['base_level_changed', \&on_possible_raise_chance]
+		);
+	} else {
+		$adding_hook = Plugins::addHooks(
+			['AI_pre',            \&on_ai_pre]
+		);
+	}
+	$status = $new_status;
+}
+
+################################################################
+#  getNextStat() is the function responsible for deciding
+#  which stat we need to raise next, according to
+#  '@stats_to_add', if we still have stats to raise, set it
+#  to '$next_stat' and return 1, if we have no stats to raise
+#  return 0.
+sub getNextStat {
+	my $amount;
+	foreach my $step (@stats_to_add) {
+		$amount = $char->{$step->{'stat'}};
+		$amount += $char->{"$step->{'stat'}_bonus"} unless $config{statsAddAuto_dontUseBonus};	
+		if ($amount < $step->{'value'}) {
+			$next_stat = $step->{'stat'};
+			return 1;
+		}
+	}
+	message "raiseStat has no more stats to raise; disabling statsAddAuto\n", 'success';
+	return 0;
+}
+
+################################################################
+#  canRaise() is the function responsible for checking if
+#  we have enough free points to raise '$next_stat'.
+sub canRaise {
+	return 1 if ($status != INACTIVE && $char->{points_free} && $char->{"points_".$next_stat} && $char->{points_free} >= $char->{"points_".$next_stat});
```

Comment:
Unless you need the values to be exactly `0` and `1`, you can just use perl's conditional result. `return` is unnecessary (but optional if you prefer it). For example:

```
sub canRaise {
	$status != INACTIVE and
	$char->{points_free} and
	$char->{"points_".$next_stat} and
	$char->{points_free} >= $char->{"points_".$next_stat};
}
```